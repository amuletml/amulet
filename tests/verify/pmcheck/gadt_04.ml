external val error : string -> 'a = "error"

type nat = Z | S of nat

type 'a × 'b <- 'a * 'b

type vect 'a 'n =
  | Nil_  : vect 'a Z
  | Cons_ : 'a × vect 'a 'n -> vect 'a (S 'n)

type fin 'n =
  | FZ : fin (S 'n)
  | FS : fin 'n -> fin (S 'n)

let rec f (vect : vect 'a 'n) (idx : fin 'n) : 'a =
  match vect, idx with
  | Nil_, _ -> error ""
  | Cons_ (x, _), FZ -> x
  | Cons_ (_, xs), FS k -> f xs k

let rec g (vect : vect 'a 'n) (idx : fin 'n) : 'a =
  match vect, idx with
  | Cons_ (x, _), FZ -> x
  | Cons_ (_, xs), FS k -> g xs k
