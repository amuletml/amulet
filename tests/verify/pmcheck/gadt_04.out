gadt_04.ml[17:5 ..17:23]: warning (W3005)
  Redundant pattern in expression
   │ 
17 │   | Nil_, _ -> error ""
   │     ^^^^^^^^^^^^^^^^^^^
  • Note: This case can never occur.
